include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+//===-- include/flang/lower/AstBuilder.h ------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_LOWER_AST_BUILDER_H_
+#define FORTRAN_LOWER_AST_BUILDER_H_
+
+#include "flang/parser/parse-tree.h"
+#include "flang/semantics/scope.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace Fortran::lower {
+namespace AST {
+
+struct Evaluation;
+struct Program;
+struct ModuleLikeUnit;
+struct FunctionLikeUnit;
+
+using ParentType =
+    std::variant<Program *, ModuleLikeUnit *, FunctionLikeUnit *, Evaluation *>;
+
+enum class CFGAnnotation {
+  None,
+  Goto,
+  CondGoto,
+  IndGoto,
+  IoSwitch,
+  Switch,
+  Iterative,
+  FirStructuredOp,
+  Return,
+  Terminate
+};
+
+/// Compiler-generated jump
+///
+/// This is used to convert implicit control-flow edges to explicit form in the
+/// decorated AST
+struct CGJump {
+  CGJump(Evaluation *to) : target{to} {}
+  Evaluation *target{nullptr};
+};
+
+/// is `A` a construct (or directive)?
+template <typename A>
+constexpr static bool isConstruct() {
+  return std::is_same_v<A, parser::AssociateConstruct> ||
+         std::is_same_v<A, parser::BlockConstruct> ||
+         std::is_same_v<A, parser::CaseConstruct> ||
+         std::is_same_v<A, parser::ChangeTeamConstruct> ||
+         std::is_same_v<A, parser::CriticalConstruct> ||
+         std::is_same_v<A, parser::DoConstruct> ||
+         std::is_same_v<A, parser::IfConstruct> ||
+         std::is_same_v<A, parser::SelectRankConstruct> ||
+         std::is_same_v<A, parser::SelectTypeConstruct> ||
+         std::is_same_v<A, parser::WhereConstruct> ||
+         std::is_same_v<A, parser::ForallConstruct> ||
+         std::is_same_v<A, parser::CompilerDirective> ||
+         std::is_same_v<A, parser::OpenMPConstruct> ||
+         std::is_same_v<A, parser::OmpEndLoopDirective>;
+}
+
+/// Function-like units can contains lists of evaluations.  These can be
+/// (simple) statements or constructs, where a construct contains its own
+/// evaluations.
+struct Evaluation {
+  using EvalVariant = std::variant<
+      // action statements
```

Comment:
1. I have split the variant as suggested.
2. As for `ParentType` I think putting reference wrapper inside variant made it more cumbersome to handle them. Since this is a member type, I it is protected it with the constructor that takes references. I made the pointer constant and added getters/visitors working on normal reference to the parse-tree to hide pointers on the user side.